src/crypto/openssl.rs

```diff
@@ -0,0 +1,941 @@
+        Self::Implementation
+    }
+}
+
+//--- Formatting
+
+impl fmt::Display for GenerateError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            Self::UnsupportedAlgorithm => "algorithm not supported",
+            Self::Implementation => "an internal error occurred",
+        })
+    }
+}
+
+//--- Error
+
+impl std::error::Error for GenerateError {}
+
+//----------- DigestContext --------------------------------------------------
+
+/// Context for computing a message digest.
+pub struct DigestContext(Hasher);
+
+impl DigestContext {
+    /// Create a new context for a specified digest type.
+    pub fn new(digest_type: DigestType) -> Self {
+        Self(
+            match digest_type {
+                DigestType::Sha1 => Hasher::new(MessageDigest::sha1()),
+                DigestType::Sha256 => Hasher::new(MessageDigest::sha256()),
+                DigestType::Sha384 => Hasher::new(MessageDigest::sha384()),
+            }
+            .expect("assume that new cannot fail"),
+        )
+    }
+
+    /// Add input to the digest computation.
+    pub fn update(&mut self, data: &[u8]) {
+        self.0
+            .update(data)
+            .expect("assume that update does not fail")
+    }
+
+    /// Finish computing the digest.
+    pub fn finish(mut self) -> Digest {
+        Digest(self.0.finish().expect("assume that finish does not fail"))
+    }
+}
+
+//----------- Digest ---------------------------------------------------------
+
+/// A message digest.
+pub struct Digest(DigestBytes);
+
+impl AsRef<[u8]> for Digest {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+//----------- PublicKey ------------------------------------------------------
+
+/// A public key for verifying a signature.
+pub enum PublicKey {
+    /// Default variant, used for RSA.
+    Default(MessageDigest, PKey<Public>, u16),
+
+    /// Variant for Ed25519 and Ed448.
+    NoDigest(PKey<Public>, u16),
+
+    /// Variant for EcDsa.
+    EcDsa(MessageDigest, EcKey<Public>, u16),
+}
+
+impl PublicKey {
+    /// Create a public key from a [`Dnskey`].
+    pub fn from_dnskey(
+        dnskey: &Dnskey<impl AsRef<[u8]>>,
+    ) -> Result<Self, AlgorithmError> {
+        let sec_alg = dnskey.algorithm();
+        match sec_alg {
+            SecAlg::RSASHA1
+            | SecAlg::RSASHA1_NSEC3_SHA1
+            | SecAlg::RSASHA256
+            | SecAlg::RSASHA512 => {
+                let (digest_algorithm, min_bytes) = match sec_alg {
+                    SecAlg::RSASHA1 | SecAlg::RSASHA1_NSEC3_SHA1 => {
+                        (MessageDigest::sha1(), 1024 / 8)
+                    }
+                    SecAlg::RSASHA256 => (MessageDigest::sha256(), 1024 / 8),
+                    SecAlg::RSASHA512 => (MessageDigest::sha512(), 1024 / 8),
+                    _ => unreachable!(),
+                };
+
+                // The key isn't available in either PEM or DER, so use the
```

Comment:
This comment doesn't apply here as no verification is being done. I assume it was copy-pasted from the original verify_signed_data() function.